### PR TITLE
Extract shared tool result assertion helper

### LINKED ISSUE #27: refactor(tests): extract assert_tool_result_contains helper and fix repeated unwrap() calls in workspace_coverage.rs
## Background

`tests/e2e_traces/workspace_coverage.rs` contains repeated inline patterns for asserting on tool results and uses `unwrap()` in multiple places where more informative error messages would be preferable. This was flagged as a deferred refactor during PR #23 (https://github.com/leynos/axinite/pull/23).

## Proposed Change

1. Extract a private helper `assert_tool_result_contains(results: &[ToolResult], tool: &str, expected: &str)` to replace the repeated inline assertion pattern.
2. Replace `unwrap()` call sites with `expect("<descriptive message>")` to improve failure diagnostics.

## Motivation

Centralising the assertion logic makes it easier to update the check if the `ToolResult` type changes, and descriptive `expect` messages reduce debugging time when a test fails in CI.

## References

- PR #23: https://github.com/leynos/axinite/pull/23

### PR DESCRIPTION
**Pull request by leynos**
## Summary

- move the repeated `assert_tool_result_contains` helper into `tests/support/assertions.rs`
- update `tests/e2e_traces/workspace_coverage.rs` to import the shared helper instead of defining it locally
- replace the remaining bare `.unwrap()` in `verify_expects` with a descriptive `.expect(...)`

## Why

`workspace_coverage.rs` duplicated inline tool-result assertion logic, which makes future changes to the result shape harder to maintain consistently. The bare `.unwrap()` in `verify_expects` also produced weaker diagnostics than the surrounding assertions when CI failures occur.

## Impact

This keeps the test behaviour unchanged while centralising the helper and improving failure messages for expectation checks.

## Validation

- `make all`
  - formatting and linting passed
  - the full test gate failed in unrelated existing `advanced_traces` cases due missing fixture files:
    - `advanced_traces::long_tool_chain`
    - `advanced_traces::user_steering`
- `cargo nextest run --features test-helpers --profile default --test e2e_traces workspace_coverage`

## References

- Closes #27
- Ref: #23

## Summary by Sourcery

Centralise shared test helpers for asserting tool results and improve expectation failure messages.

Enhancements:
- Replace a bare unwrap in verify_expects with a descriptive expect message to improve diagnostics.

Tests:
- Extract the tool result substring assertion into a shared helper in tests/support/assertions.rs and update workspace_coverage tests to use it.

## Summary by Sourcery

Extract a shared test helper for asserting tool result contents and improve diagnostics in expectation verification.

Enhancements:
- Move the tool result substring assertion into the shared test assertions module for reuse across tests.
- Replace a bare unwrap in verify_expects with a descriptive expect message to provide clearer failure diagnostics.

Tests:
- Update workspace_coverage end-to-end trace tests to use the shared tool result assertion helper instead of a local implementation.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The new `assert_tool_result_contains` helper performs case-sensitive `contains` checks, whereas `verify_expects` uses case-insensitive matching via `to_lowercase`; consider aligning their behavior or documenting the intentional difference so callers know what to expect.

## Individual Comments

### Comment 1
<location path="tests/support/assertions.rs" line_range="85-105" />
<code_context>
     );
 }

+/// Assert that at least one result for `tool_name` contains an expected substring.
+pub fn assert_tool_result_contains(
+    results: &[(String, String)],
+    tool_name: &str,
+    expected_substrings: &[&str],
+) {
+    let matched: Vec<_> = results
+        .iter()
+        .filter(|(name, _)| name == tool_name)
+        .collect();
+    assert!(
+        !matched.is_empty(),
+        "Expected at least one result for tool '{tool_name}'"
+    );
+    assert!(
+        matched
+            .iter()
+            .any(|(_, preview)| expected_substrings.iter().any(|s| preview.contains(s))),
+        "No result for '{tool_name}' contained any of {expected_substrings:?}: {matched:?}"
+    );
+}
</code_context>
<issue_to_address>
**suggestion:** Clarify or guard against the case where `expected_substrings` is empty

If `expected_substrings` is empty, this will always fail the second assertion with `matched` populated but `any(..)` returning `false`. While that’s logically fine, the failure message about not containing "any of {expected_substrings}" is confusing when the set is empty. Consider either asserting up front that `expected_substrings` is non-empty, or documenting that requirement in the function comment so failures are clearer to future readers.

```suggestion
/// Assert that at least one result for `tool_name` contains an expected substring.
///
/// # Panics
///
/// Panics if:
/// - `expected_substrings` is empty,
/// - there are no results for `tool_name`, or
/// - none of the matching results contain any of the expected substrings.
pub fn assert_tool_result_contains(
    results: &[(String, String)],
    tool_name: &str,
    expected_substrings: &[&str],
) {
    assert!(
        !expected_substrings.is_empty(),
        "expected_substrings must not be empty when asserting tool results for '{tool_name}'"
    );

    let matched: Vec<_> = results
        .iter()
        .filter(|(name, _)| name == tool_name)
        .collect();
    assert!(
        !matched.is_empty(),
        "Expected at least one result for tool '{tool_name}'"
    );
    assert!(
        matched
            .iter()
            .any(|(_, preview)| expected_substrings.iter().any(|s| preview.contains(s))),
        "No result for '{tool_name}' contained any of {expected_substrings:?}: {matched:?}"
    );
}
```
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Developer Documentation | ⚠️ Warning | The PR adds a new public function `assert_tool_result_contains` without inline Rust doc comments and includes inadequate markdown documentation with incorrect heading hierarchy. | Add comprehensive Rust doc comments to `assert_tool_result_contains` documenting purpose, parameters, and behaviour. Correct the "Shared test assertions" heading from level-3 to level-4 in developers-guide.md. |
| Module-Level Documentation | ⚠️ Warning | The module `tests/support_unit_tests/assertions_tests.rs` lacks adequate module-level documentation with only a bare label docstring. | Expand the docstring to describe unit tests for assertion helpers, covering case-insensitive matching, tool filtering, panic diagnostics, and edge cases. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tests/support_unit_tests/assertions_tests.rs](https://github.com/leynos/axinite/pull/162/files/5a277246555cce522e1478088b92c68ce769b7c7#diff-20c16c71bd54e8c8edaef58a812a9dd53b2ef27ddeeb0b766505b3dff1cc5abb)

Comment on lines +122 to +134

~~~rust
fn tool_result_contains_panics_when_tool_is_missing() {
    let results = vec![("memory_search".to_string(), "Alpha/Beta".to_string())];

    let panic = catch_unwind(AssertUnwindSafe(|| {
        assert_tool_result_contains(&results, "memory_tree", &["alpha"]);
    }))
    .expect_err("helper should fail when the tool is missing");

    assert_eq!(
        panic_message(panic),
        "Expected at least one result for tool 'memory_tree'"
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=162&biomarker=Code+Duplication&filename=tests%2Fsupport_unit_tests%2Fassertions_tests.rs&method=)
The module contains 2 functions with similar structure: tool_result_contains_panics_when_tool_is_missing,tool_result_contains_rejects_empty_expected_substrings

File: tests/support/assertions.rs [hunk 1]
```diff
@@ -204,10 +246,34 @@ pub fn verify_expects(
...
+    #[test]
+    #[should_panic(
+        expected = "expected_substrings must not be empty when asserting tool results for 'memory_tree'"
+    )]
+    fn assert_tool_result_contains_rejects_empty_expected_substrings() {
+        let results = vec![("memory_tree".to_string(), "Alpha/Beta".to_string())];
+
+        assert_tool_result_contains(&results, "memory_tree", &[]);
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/assertions.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/support/assertions.rs [hunk 2]
```diff
@@ -82,6 +82,48 @@ pub fn assert_tool_succeeded(completed: &[(String, bool)], tool_name: &str) {
...
+) {
+    assert!(
+        !expected_substrings.is_empty(),
+        "expected_substrings must not be empty when asserting tool results for '{tool_name}'"
+    );
+
+    let lower_expected: Vec<String> = expected_substrings
+        .iter()
+        .map(|substring| substring.to_lowercase())
+        .collect();
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/assertions.rs:118]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: docs/developers-guide.md [hunk 3]
```diff
@@ -257,6 +257,28 @@ Tests:
...
+  status checks
+- `assert_tool_result_contains` for case-insensitive preview matching against
+  the captured `(tool_name, preview)` results emitted by the test rig
+- `verify_expects` for applying `TraceExpects` fixtures to a captured test run
+
+Keep broad behavioural coverage in the E2E trace fixtures, and place focused
+helper regression tests in `tests/support_unit_tests/assertions_tests.rs` so
+they run once instead of being duplicated across every E2E binary.
+
 #### AppBuilderFlags
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | Snapshot tests for panic diagnostics (lines 182-209) and parameterised rstest cases for duplicate guard tests (lines 155-179) remain unimplemented. | Implement snapshot testing using insta::assert\_snapshot! for panic messages and collapse duplicate tests into single parameterised #\[rstest\] case. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[tests/support_unit_tests/assertions_tests.rs](https://github.com/leynos/axinite/pull/162/files/d7d55c91075c062e2f8a505a9459a968f47d46e6#diff-20c16c71bd54e8c8edaef58a812a9dd53b2ef27ddeeb0b766505b3dff1cc5abb)

Comment on lines +196 to +208

~~~rust
fn verify_expects_reports_missing_substring_with_preview() {
    let expects = TraceExpects {
        tool_results_contain: HashMap::from([("memory_tree".to_string(), "alpha".to_string())]),
        ..TraceExpects::default()
    };
    let results = vec![("memory_tree".to_string(), "Gamma/Delta".to_string())];

    let message = capture_panic_message(AssertUnwindSafe(|| {
        verify_expects(&expects, &[], &[], &[], &results, "turn 0");
    }));

    assert_snapshot!(message);
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/77987/delta?repo-id=1501128&review-id=162&biomarker=Code+Duplication&filename=tests%2Fsupport_unit_tests%2Fassertions_tests.rs&method=)
The module contains 2 functions with similar structure: verify_expects_reports_missing_substring_with_preview,verify_expects_reports_missing_tool_result_with_label

File: tests/support/assertions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leynos) [reply tests/support/assertions.rs:118]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
